---
layout: default
title: Liu Lab
subtitle: Xiaole Shirley Liu Lab
use-site-title: true
---

<div class="jumbotron">
    <div class="container">
        <div class="row">
            <div class="col-lg-3">
                <p>
                    <img src="/img/people/shirley_liu.png" width="230" height="230">
                </p>
            </div>
            <div class="col-lg-9">
                <p>X. Shirley Liu received a Ph.D. in Biomedical Informatics and a Ph.D. minor in Computer Science from Stanford University in 2002.
She has been a Professor with the Department of Biostatistics and Computational Biology at the Dana-Farber Cancer Institute and Harvard T.H. Chan School of Public Health until 2022.
Shirley has been a computational biologist with expertise in cancer epigenetics and cancer immunology.
Her work has been focused on algorithm development and data integration modeling for translational cancer research. 
Her laboratory developed widely used algorithms and tools for transcription factor motif finding, ChIP-chip/seq, chromatin accessibility profiles, CRISPR screen analyses, and tumor immune characterization. 
She left DFCI / Harvard to join GV20 Oncotherapy on 2/1/2022.

                </p>
            </div>
        </div>
    </div>
</div>

<div class="container">
    <div class="row">
        <h2>Research interests</h2>
        <p>Our research focuses on algorithm development and integrative mining from high-throughput data to understand gene regulation in cancer biology.
            We have developed a number of widely used algorithms for transcription factor motif finding, ChIP-chip / ChIP-seq / DNase-seq / CRISPR Screen data analysis.
            Through integrating genome-wide transcription factor binding, chromatin dynamics, gene expression profiles, and chemical and functional screens, we try to model the specificity and function of transcription factors, chromatin regulators, RNA binding proteins, kinases, and lncRNAs in tumor development, progression, drug response and resistance.
        </p>
    </div>
    <div class="row">
        <h3>Cancer Epigenetics</h3>
        <div class="col-lg-3">
            <img src="/img/research/Cistrome.png">
        </div>
        <div class="col-lg-9">
            <p>We have been developing algorithms (MACS, Cistrome, NPS, BETA) to facilitate the analysis of epigenomic data and use integrative modeling approaches to study genomic transcriptional and epigenetic gene regulatory mechanisms underlying tumorigenesis and progression.
                We are developing new methods to utilize the abundant public ChIP-seq data to infer functional enhancers that regulate the expression, and understand the targets and specificity of epigenetic drugs in cancers.
            </p>
        </div>
    </div>
    <div class="row">
        <h3>CRISPR screens</h3>
        <div class="col-lg-3">
            <img src="/img/research/CRISPR.png">
        </div>
        <div class="col-lg-9">
            <p>
                We are developing the computational methods for the design (SSC), analysis (MAGeCK), hit prioritization (NEST), and visualization (VISPR) of genome-wide CRISPR screens.
                We are also using this technology to identify key genes in breast and prostate tumor progression and drug resistance.
                We also develop CRISPR screen platforms to understand the functions of enhancers and long-noncoding RNAs, and identify synthetic lethal gene pairs in cancer that leads to optimized cancer precision medicine.
            </p>
        </div>
    </div>
    <div class="row">
        <h3>Cancer Immunology</h3>
        <div class="col-lg-3">
            <img src="/img/research/TIMER.png">
        </div>
        <p>
            We are developing novel computational algorithms to systematically deconvolve the abundance of different immune cell types.
            Applying the deconvolution method on over 10K tumor samples from TCGA identified widespread clinical associations with tumor infiltrating immune cells.
            Our investigations on cancer vaccine and checkpoint blockade targets suggest that systematic analyses of tumor immunity have the potential to inform effective cancer immunotherapies.
        </p>
    </div>
    <div class="row">
        <h3>Single Cell</h3>
        <div class="col-lg-3">
            <img src="/img/research/MAESTRO.png">
        </div>
        <p>
            We are developing computational methods (<a href="http://github.com/liulab-dfci/MAESTRO">MAESTRO</a>) for integrative analysis of single-cell RNA-seq (scRNA-seq) and ATAC-seq (scATAC-seq) data. By integrating the public transcription factor (TF) ChIP-seq datasets in Cistrome DB, we are unraveling the regulatory network with tumor microenvironment and identifying key regulators between different immune cells. We are also developing novel computational algorithms for spatial transcriptomic data analysis.
        </p>
    </div>

</div>

<!--<div class="posts-list">-->
  <!--{% for post in paginator.posts %}-->
  <!--<article class="post-preview">-->
    <!--<a href="{{ post.url | relative_url }}">-->
	  <!--<h2 class="post-title">{{ post.title }}</h2>-->

	  <!--{% if post.subtitle %}-->
	  <!--<h3 class="post-subtitle">-->
	    <!--{{ post.subtitle }}-->
	  <!--</h3>-->
	  <!--{% endif %}-->
    <!--</a>-->

    <!--<p class="post-meta">-->
      <!--Posted on {{ post.date | date: site.date_format }}-->
    <!--</p>-->

    <!--<div class="post-entry-container">-->
      <!--{% if post.image %}-->
      <!--<div class="post-image">-->
        <!--<a href="{{ post.url | relative_url }}">-->
          <!--<img src="{{ post.image | relative_url }}">-->
        <!--</a>-->
      <!--</div>-->
      <!--{% endif %}-->
      <!--<div class="post-entry">-->
        <!--{{ post.excerpt | strip_html | xml_escape | truncatewords: site.excerpt_length }}-->
        <!--{% assign excerpt_word_count = post.excerpt | number_of_words %}-->
        <!--{% if post.content != post.excerpt or excerpt_word_count > site.excerpt_length %}-->
          <!--<a href="{{ post.url | relative_url }}" class="post-read-more">[Read&nbsp;More]</a>-->
        <!--{% endif %}-->
      <!--</div>-->
    <!--</div>-->

    <!--{% if post.tags.size > 0 %}-->
    <!--<div class="blog-tags">-->
      <!--Tags:-->
      <!--{% if site.link-tags %}-->
      <!--{% for tag in post.tags %}-->
      <!--<a href="{{ '/tags' | relative_url }}#{{- tag -}}">{{- tag -}}</a>-->
      <!--{% endfor %}-->
      <!--{% else %}-->
        <!--{{ post.tags | join: ", " }}-->
      <!--{% endif %}-->
    <!--</div>-->
    <!--{% endif %}-->

   <!--</article>-->
  <!--{% endfor %}-->
<!--</div>-->

<!--{% if paginator.total_pages > 1 %}-->
<!--<ul class="pager main-pager">-->
  <!--{% if paginator.previous_page %}-->
  <!--<li class="previous">-->
    <!--<a href="{{ paginator.previous_page_path | relative_url }}">&larr; Newer Posts</a>-->
  <!--</li>-->
  <!--{% endif %}-->
  <!--{% if paginator.next_page %}-->
  <!--<li class="next">-->
    <!--<a href="{{ paginator.next_page_path | relative_url }}">Older Posts &rarr;</a>-->
  <!--</li>-->
  <!--{% endif %}-->
<!--</ul>-->
<!--{% endif %}-->
